wrappers/rust/icicle-core/src/msm/tests.rs

```diff
@@ -19,61 +22,73 @@ where
+            //      3) Perform all operations (without changing device on the thread)
+            //      4) If necessary - export results to parent host thread
 
-        let mut cfg = get_default_msm_config::<C>();
-        cfg.ctx
-            .stream = &stream;
-        cfg.is_async = true;
-        cfg.are_scalars_montgomery_form = true;
-        msm(&scalars_d, &HostOrDeviceSlice::on_host(points), &cfg, &mut msm_results).unwrap();
-        // need to make sure that scalars_d weren't mutated by the previous call
-        let mut scalars_mont_after = vec![C::ScalarField::zero(); test_size];
-        scalars_d
-            .copy_to_host_async(&mut scalars_mont_after, &stream)
-            .unwrap();
-        assert_eq!(scalars_mont, scalars_mont_after);
+            set_device(device_id).unwrap();
+            let test_sizes = [4, 8, 16, 32, 64, 128, 256, 1000, 1 << 18];
+            let mut msm_results = HostOrDeviceSlice::cuda_malloc(1).unwrap();
+            for test_size in test_sizes {
+                let points = C::generate_random_affine_points(test_size);
+                let scalars = <C::ScalarField as FieldImpl>::Config::generate_random(test_size);
+                let points_ark: Vec<_> = points
+                    .iter()
+                    .map(|x| x.to_ark())
+                    .collect();
+                let scalars_ark: Vec<_> = scalars
+                    .iter()
+                    .map(|x| x.to_ark())
+                    .collect();
+                // if we simply transmute arkworks types, we'll get scalars or points in Montgomery format
+                // (just beware the possible extra flag in affine point types, can't transmute ark Affine because of that)
+                let scalars_mont = unsafe { &*(&scalars_ark[..] as *const _ as *const [C::ScalarField]) };
 
-        let mut msm_host_result = vec![Projective::<C>::zero(); 1];
-        msm_results
-            .copy_to_host(&mut msm_host_result[..])
-            .unwrap();
-        stream
-            .synchronize()
-            .unwrap();
-        stream
-            .destroy()
-            .unwrap();
+                let mut scalars_d = HostOrDeviceSlice::cuda_malloc(test_size).unwrap();
+                let stream = CudaStream::create().unwrap();
+                scalars_d
+                    .copy_from_host_async(&scalars_mont, &stream)
+                    .unwrap();
 
-        let msm_result_ark: ark_ec::models::short_weierstrass::Projective<C::ArkSWConfig> =
-            VariableBaseMSM::msm(&points_ark, &scalars_ark).unwrap();
-        let msm_res_affine: ark_ec::short_weierstrass::Affine<C::ArkSWConfig> = msm_host_result[0]
-            .to_ark()
-            .into();
-        assert!(msm_res_affine.is_on_curve());
-        assert_eq!(msm_host_result[0].to_ark(), msm_result_ark);
-    }
+                let mut cfg = get_default_msm_config::<C>();
```

Comment:
Related to the previous comment: doesn't this create config on the first device only?

Reply:
fixed